codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customizations/AddValidationExceptionToConstrainedOperationsTest.kt

```diff
@@ -0,0 +1,133 @@
+            output: SampleInputOutput
+            input: SampleInputOutput
+            errors: [ValidationException, ErrorWithMemberConstraint]
+        }
+        @http(uri: "/anOperationWithoutValidation", method: "POST")
+        operation SampleOperationWithoutValidation {
+            output: SampleInputOutput
+            input: SampleInputOutput
+            errors: []
+        }
+        structure SampleInputOutput {
+            constrainedInteger : RangedInteger
+            @range(min: 2, max:100)
+            constrainedMemberInteger : RangedInteger
+            patternString : PatternString
+        }
+        @pattern("^[a-m]+${'$'}")
+        string PatternString
+        @range(min: 0, max:1000)
+        integer RangedInteger
+
+        @error("server")
+        structure ErrorWithMemberConstraint {
+            @range(min: 100, max: 999)
+            statusCode: Integer
+        }
+        """.asSmithyModel(smithyVersion = "2")
+
+    private val testModelWithoutValidationExceptionImported =
+        """
+        namespace test
+
+        use aws.protocols#restJson1
+        use aws.api#data
+
+        @restJson1
+        service ConstrainedService {
+            operations: [SampleOperationWithoutValidation]
+        }
+        @http(uri: "/anOperationWithoutValidation", method: "POST")
+        operation SampleOperationWithoutValidation {
+            output: SampleInputOutput
+            input: SampleInputOutput
+            errors: []
+        }
+        structure SampleInputOutput {
+            constrainedInteger : RangedInteger
+            @range(min: 2, max:100)
+            constrainedMemberInteger : RangedInteger
+            patternString : PatternString
+        }
+        @pattern("^[a-m]+${'$'}")
+        string PatternString
+        @range(min: 0, max:1000)
+        integer RangedInteger
+
+        @error("server")
+        structure ErrorWithMemberConstraint {
+            @range(min: 100, max: 999)
+            statusCode: Integer
+        }
+        """.asSmithyModel(smithyVersion = "2")
+
+    @Test
+    fun `without setting the codegen flag, the model should fail to compile`() {
+        assertThrows<CodegenException> {
+            serverIntegrationTest(
+                testModelWithValidationExceptionImported,
+                IntegrationTestParams(),
+            )
+        }
+    }
+
+    @Test
+    fun `operations that do not have ValidationException will automatically have one added to them`() {
+        serverIntegrationTest(
+            testModelWithValidationExceptionImported,
+            IntegrationTestParams(
+                additionalSettings =
+                    ServerAdditionalSettings.builder()
+                        .addValidationExceptionToConstrainedOperations()
+                        .toObjectNode(),
+            ),
+        )
+    }
+
+    @Test
+    fun `the model, which does not have Smithy's ValidationException imported, should work when the codegen parameter is set`() {
```

Comment:
I don't think this test is necessary. The fact that a shape is `use`d or not in the model at the source-level is irrelevant once smithy-rs runs, which gets an already parsed `Model`.

Reply:
Oh, that’s nice. But how does `smithy.framework#ValidationException` get introduced into the shapes in the model, especially when we’re just adding a ShapeID to the transformation and not an in-memory shape? Why doesn’t it complain that the shape is unknown?